File: server/lib/openthink_backend/messages/message.ex
defmodule OpenthinkBackend.Message do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:message_id, :id, autogenerate: true}

  schema "messages" do
    field :created_by, :integer
    field :body, :string
    field :room_id, :integer
    field :post_id, :integer
    field :delta, :string
    field :edited, :boolean
    field :pinned, :boolean
    field :created_at, :utc_datetime
    field :updated_at, :utc_datetime
    field :deleted, :boolean
    field :replying_to, :integer
    has_one :user, OpenthinkBackend.User, foreign_key: :user_id, references: :created_by
    has_many :reactions, OpenthinkBackend.Reaction, foreign_key: :message_id
    has_many :mentions, OpenthinkBackend.Mention, foreign_key: :message_id
  end

  def changeset(message, params \\ %{}) do
    message
    |> cast(params, [:body, :created_by, :room_id, :post_id, :delta, :replying_to])
    |> validate_required([:body, :created_by])
  end

  def delete_changeset(message, params \\ %{}) do
    message
    |> cast(params, [:deleted])
    |> validate_required([:deleted])
  end

  def update_message_changeset(message, attrs \\ %{}) do
    message
    |> cast(attrs, [
      :body
      ])
    |> validate_required([
      :body
    ])
  end

  def pin_message_changeset(message, attrs \\ %{}) do
    message
    |> cast(attrs, [
      :pinned
      ])
    |> validate_required([
      :pinned
    ])
  end
end

File: client/src/types/message.ts
import { user } from "./user";

export interface message {
    message_id?: number;
    created_by?: number;
    body?: string;
    room_id?: number;
    edited?: boolean;
    pinned?: boolean;
    created_at?: string;
    updated_at?: Date;
    files?: any;
    reactions?: reaction[]
    firstname?: string,
    lastname?: string,
    profilepic?: string,
    createdAt?: Date,
    user?: user,
    createdBy?: Date,
    messageId?: number,
    delta?: any,
    mentions?: mention[],
    id?: string,
    space_id?: number
}

export interface reaction {
    emoji: string;
    unified: string;
    user_id: number;
    message_id: number;
    name: string;
    users: user[];
}

export interface mention {
    user_id: number;
    message_id?: number;
    post_id?: number;
}


export interface room {
    id?: string;
    room_id?: number;
    name?: string;
    otherUser?: user;
    space_id?: number;
    spaceId?: number;
    archived?: boolean;
    description?: string;
    dm?: boolean;
    created_by?: number;
    created_at?: Date;
    profilepic?: string;
    users?: user[];
    scrollPos?: number;
    offset?: number;
    unread?: boolean | null;
    unreadNum?: number | null;
    hasMore?: boolean;
    scrolledUp?: boolean;
    private?: boolean;
    muted?: boolean;
    notification_settings?: number;
    muted_until?: Date | null,
    roomId?: number,
    visibility?: string,
    type?: string,
}

export interface role {
    role_id: number,
    space_id: number,
    name: string,
    color: string,
    description: string,
    moderator: boolean,
}

export interface category {
    name: string;
    private?: boolean;
    users: (user | role)[]
}